domains/pallets/ethereum-beacon-client/src/lib.rs

```diff
@@ -0,0 +1,1093 @@
+        NextSyncCommitteeMismatch,
+        /// Finalized header's sync committee period does not match store's current period
+        FinalizedPeriodMismatch,
+        /// Unable to store non sequential sync committee update
+        NonSequentialSyncCommitteeUpdate,
+    }
+
+    #[pallet::hooks]
+    impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> {}
+
+    /// Historical execution headers
+    #[pallet::storage]
+    pub(super) type ExecutionHeaders<T: Config> =
+        CountedStorageMap<_, Identity, H256, ExecutionHeaderOf, OptionQuery>;
+
+    /// Mapping of count -> Execution header hash. Used to prune older headers
+    #[pallet::storage]
+    pub(super) type ExecutionHeadersMapping<T: Config> =
+        StorageMap<_, Identity, u64, H256, ValueQuery>;
+
+    /// Historical sync committees till the sync committee of latest finalized header.
+    #[pallet::storage]
+    pub(super) type SyncCommittees<T: Config> =
+        CountedStorageMap<_, Identity, u64, SyncCommitteeOf, ValueQuery>;
+
+    /// Genesis validators root
+    #[pallet::storage]
+    pub(super) type GenesisValidatorsRoot<T: Config> = StorageValue<_, H256, ValueQuery>;
+
+    /// Latest finalized header observed
+    #[pallet::storage]
+    pub(super) type LatestFinalizedHeaderState<T: Config> =
+        StorageValue<_, FinalizedHeaderState, ValueQuery>;
+
+    /// Latest execution header observed
+    #[pallet::storage]
+    pub(super) type LatestExecutionHeaderState<T: Config> =
+        StorageValue<_, ExecutionHeaderState, ValueQuery>;
+
+    /// Latest sync committee period observed
+    #[pallet::storage]
+    pub(super) type LatestSyncCommitteePeriod<T: Config> = StorageValue<_, u64, ValueQuery>;
+
+    /// Boolean signifying whether the bridge is blocked or not
+    #[pallet::storage]
+    pub(super) type Blocked<T: Config> = StorageValue<_, bool, ValueQuery>;
+
+    #[pallet::genesis_config]
+    #[derive(Default)]
+    pub struct GenesisConfig {
+        pub initial_sync: Option<InitialSyncOf>,
+    }
+
+    #[cfg(feature = "std")]
+    #[pallet::genesis_build]
+    impl<T: Config> GenesisBuild<T> for GenesisConfig {
+        fn build(&self) {
+            log::info!(
+                target: "ethereum-beacon-client",
+                "💫 Sync committee size is: {}",
+                config::SyncCommitteeSize::get()
+            );
+
+            if let Some(initial_sync) = self.initial_sync.clone() {
+                Pallet::<T>::initial_sync(initial_sync).expect("Genesis sync cannot fail");
+            }
+        }
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+        /// Updates the light client state
+        #[pallet::weight(T::WeightInfo::sync_committee_period_update())]
+        #[pallet::call_index(0)]
+        pub fn light_client_update(
+            origin: OriginFor<T>,
+            light_client_update: LightClientUpdateOf,
+        ) -> DispatchResult {
+            let _sender = ensure_signed(origin)?;
```

Comment:
I guess this and `import_execution_header` below will be changed to the relayer only in the future?

Reply:
I recall somewhere in the design doc the relayer was mentioned, thus had this question, I could be wrong.

Reply:
@vedhavyas @liuchengxu This will depend upon whether we want relayer incentivization in near future or not. My personal opinion is that we do not need to implement it now and anybody who can pay the tx fee can submit the *untrusted* data to the light client.